crates/subspace-farmer/src/piece_cache.rs

```diff
@@ -0,0 +1,666 @@
+                    }
+                    return;
+                }
+
+                warn!(
+                    %piece_index,
+                    "Should have inserted piece into cache, but it didn't happen, this is an \
+                    implementation bug"
+                );
+            }
+        };
+    }
+}
+
+/// Piece cache that aggregates caches of multiple disks
+#[derive(Debug, Clone)]
+pub struct PieceCache {
+    /// Individual disk caches where pieces are stored
+    caches: Arc<RwLock<Vec<DiskPieceCacheState>>>,
+    // We do not want to increase capacity unnecessarily on clone
+    worker_sender: mpsc::Sender<WorkerCommand>,
+}
+
+impl PieceCache {
+    /// Create new piece cache instance and corresponding worker.
+    ///
+    /// NOTE: Returned future is async, but does blocking operations and should be running in
+    /// dedicated thread.
+    pub fn new<NC>(node_client: NC, peer_id: PeerId) -> (Self, CacheWorker<NC>)
+    where
+        NC: NodeClient,
+    {
+        let caches = Arc::default();
+        let (worker_sender, worker_receiver) = mpsc::channel(WORKER_CHANNEL_CAPACITY);
+
+        let instance = Self {
+            caches: Arc::clone(&caches),
+            worker_sender,
+        };
+        let worker = CacheWorker {
+            peer_id,
+            node_client,
+            caches,
+            worker_receiver: Some(worker_receiver),
+        };
+
+        (instance, worker)
+    }
+
+    /// Check whether piece exists in cache
+    pub fn contains_piece(&self, key: RecordKey) -> bool {
+        // It is okay to take read lock here, writes locks are very infrequent and very short
+        for cache in self.caches.read().iter() {
+            if cache.stored_pieces.contains_key(&key) {
+                return true;
+            };
+        }
+
+        false
+    }
+
+    /// Get piece from cache
+    pub async fn get_piece(&self, key: RecordKey) -> Option<Piece> {
+        let caches = Arc::clone(&self.caches);
+
+        // TODO: Join on drop for this async task
+        let maybe_piece_fut = tokio::task::spawn_blocking({
+            let key = key.clone();
+            let worker_sender = self.worker_sender.clone();
+
+            move || {
+                for (disk_farm_index, cache) in caches.read().iter().enumerate() {
+                    let Some(&offset) = cache.stored_pieces.get(&key) else {
+                        continue;
+                    };
+                    match cache.backend.read_piece(offset) {
+                        Ok(maybe_piece) => {
+                            return maybe_piece;
+                        }
+                        Err(error) => {
+                            error!(
+                                %error,
+                                %disk_farm_index,
+                                ?key,
+                                %offset,
+                                "Error while reading piece from cache, might be a disk corruption"
```

Comment:
or any other disk IO error (like insufficient permissions).

Reply:
No, we use memory-mapped files, when this read happens we work with essentially `&[u8]` as far as the function call is concerned.